etk-asm/src/asm.rs

```diff
@@ -59,68 +89,121 @@ impl Assembler {
         if self.pending.is_empty() {
-            self.push_ready(op)?;
+            self.push_ready(node)
         } else {
-            self.push_pending(op)?;
+            self.push_pending(node)
         }
-
-        self.code_len += 1 + specifier.extra_len();
-        Ok(self.ready.len())
     }
 
-    fn push_ready(&mut self, mut op: Op) -> Result<(), Error> {
-        if let Some(label) = op.immediate_label() {
-            match self.labels.get(label) {
-                Some(addr) => op = op.realize(*addr)?,
-                None => {
-                    self.pending.push_back(op);
-                    return Ok(());
+    fn push_ready(&mut self, mut node: Node) -> Result<(), Error> {
+        match node.clone() {
+            Node::Op(op) => {
+                if let Some(label) = op.immediate_label() {
+                    match self.labels.get(label) {
+                        Some(addr) => {
+                            node = op.realize(*addr)?.into();
+                        }
+                        None => {
+                            self.pending.push_back(node.clone());
+                            return Ok(());
+                        }
+                    }
                 }
-            }
-        }
 
-        op.assemble(&mut self.ready);
+                node.assemble(&mut self.ready);
 
-        Ok(())
+                Ok(())
+            }
+            Node::Raw(raw) => {
+                self.ready.extend(raw);
+                Ok(())
+            }
+            _ => unimplemented!(),
+        }
     }
 
-    fn push_pending(&mut self, op: Op) -> Result<(), Error> {
-        self.pending.push_back(op);
+    fn push_pending(&mut self, node: Node) -> Result<(), Error> {
+        self.pending.push_back(node);
 
         while let Some(next) = self.pending.front() {
             let mut address = None;
 
-            if let Some(label) = next.immediate_label() {
-                match self.labels.get(label) {
-                    Some(addr) => address = Some(*addr),
-                    None => break,
+            match next {
+                Node::Op(op) => {
+                    if let Some(label) = op.immediate_label() {
+                        match self.labels.get(label) {
+                            Some(addr) => address = Some(*addr),
+                            None => break,
+                        }
+                    }
+
+                    let popped = match address {
+                        Some(s) => {
+                            // Don't modify `self.pending` if realize returns an error.
+                            let realized = op.realize(s)?;
+                            self.pending.pop_front();
+                            Node::Op(realized)
+                        }
+                        None => self.pending.pop_front().unwrap(),
+                    };
+
+                    popped.assemble(&mut self.ready);
                 }
+                _ => unimplemented!(),
```

Comment:
Why does `push_pending` take a `Node` if anything other than an `Op` is unimplemented?

Reply:
I should rewrite that portion to be `if let Node::Op(op) = next`... but `self.pending.push_back(node)` occurs for things other than op so it should include `node`.